### PR TITLE
Add testbed mode

### PR DESCRIPTION
**Pull request by ximon18**
Adds support for a new `testbed_enabled` boolean flag to configure Krill to permit child/parent CA/repo RFC-8183 relationships to be established by *anonymous* clients with a CA called `testbed` and enables a `/#/testbed` UI page for doing so visually ala other RPKI testbeds such as the APNIC RPKI testbed. Note: the `testbed` CA must be created manually, it is **not** created as part of this PR Implements the s/w parts of issue #306.

### CODE REVIEW

**Comment by ximon18** [comment]
Note: merge to master should probably wait on merge of the Lagosta `testbed-poc` branch to Lagosta master first. That branch is included (in release compiled form) in this Krill branch.

**Comment by timbru** [comment]
> Note: merge to master should probably wait on merge of the Lagosta `testbed-poc` branch to Lagosta master first. That branch is included (in release compiled form) in this Krill branch.

I don't mind this.. we are merging to a dev branch and can update lagosta again later.

Also the failing test may be due to the issue with the message queue scheduling fixed in '9b46f5ab60b49485d6e18e6770e0478c73e52b85'

So, it's probably fine to mere.

**Comment by ximon18** [comment]
Thanks @timbru. FYI this work currently lacks any modification of or addition to the automated test suite. I should probably extend it at least to check that the disabling of authorisation checks only happens as and when it should, right?

**Comment by timbru** [comment]
> Thanks @timbru. FYI this work currently lacks any modification of or addition to the automated test suite. I should probably extend it at least to check that the disabling of authorisation checks only happens as and when it should, right?

That sounds like a really good idea :)

Other than that it would also be nice to have an end to end test for this so we can catch regressions.

**Comment by ximon18** [comment]
The CI build failure ([`reached the type-length limit while instantiating`](https://github.com/NLnetLabs/krill/pull/308/checks?check_run_id=1020520324#step:4:663)) is only on Rust beta, not on Rust stable, so will ignore for now and assume it will soon solve itself.

File: src/daemon/http/server.rs [hunk 1]
```diff
@@ -1285,6 +1261,61 @@ async fn api_ca_rta_oneoff(req: Request, path: &mut RequestPath, ca: Handle) ->
     }
 }
 
+//------------ Support acting as a testbed -------------------------------------
+
+async fn testbed(req: Request) -> RoutingResult {
+    if !CONFIG.testbed_enabled {
+        render_not_found(req).await
```
Comment (timbru) [inline src/daemon/http/server.rs:1268]: Use `Err(req)` so that the req gets returned and routing continues.

File: src/daemon/http/server.rs [hunk 2]
```diff
@@ -1285,6 +1261,61 @@ async fn api_ca_rta_oneoff(req: Request, path: &mut RequestPath, ca: Handle) ->
...
+    } else {
+        let mut path = req.path().clone();
+        match path.next() {
+            Some("enabled")    => testbed_enabled(req).await,
+            Some("children")   => testbed_children(req, &mut path).await,
+            Some("publishers") => testbed_publishers(req, &mut path).await,
+            _                  => render_not_found(req).await,
```
Comment (timbru) [inline src/daemon/http/server.rs:1275]: The api uses `render_unknown_method()` whenever the URI maps to a specific part of the API, but there is no method to map to. As opposed to `render_unknown_resource()` when the path is okay but the actual resource at that location cannot be found (e.g. a CA does not exist). The first results in a 400 response, the second in a 404 with a json error, whereas `render_not_found(req)` results in a 404 with HTML which is intended to be browser user facing.

Whether this is all the best way of doing things is of course debatable, but we should still use `render_unknown_method()` here to be consistent.

**Comment by ximon18** [comment]
TODO: update to use Lagosta master as the Lagosta testbed-poc branch has been merged.

**Comment by ximon18** [comment]
Agreed with @timbru to merge this branch.

**Pull request merged**